File: skellig-test-step-runner/src/test/java/org/skellig/teststep/runner/DefaultTestStepRunnerTest.java
package org.skellig.teststep.runner;


import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.skellig.teststep.processing.exception.TestStepProcessingException;
import org.skellig.teststep.processing.model.TestStep;
import org.skellig.teststep.processing.model.factory.TestStepFactory;
import org.skellig.teststep.processing.processor.TestStepProcessor;
import org.skellig.teststep.processing.state.TestScenarioState;
import org.skellig.teststep.reader.TestStepReader;

import java.net.URISyntaxException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Collections;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyZeroInteractions;
import static org.mockito.Mockito.when;

@DisplayName("Run test step")
class DefaultTestStepRunnerTest {

    private TestStepRunner testStepRunner;
    private TestStepProcessor testStepProcessor;
    private TestStepReader testStepReader;
    private TestStep testStep;
    private TestStepFactory testStepFactory;
    private TestScenarioState testScenarioState;

    @BeforeEach
    void setUp() {
        testStepProcessor = mock(TestStepProcessor.class);
        testStepReader = mock(TestStepReader.class);
        testStepFactory = mock(TestStepFactory.class);
        testScenarioState = mock(TestScenarioState.class);
    }

    @Test
    @DisplayName("When no parameters extracted from name")
    void testRunTestStepWithoutParameters() throws URISyntaxException {
        String testStepName = "test1";
        initializeTestSteps(testStepName, Collections.emptyMap());
        initializeTestStepRunner();

        testStepRunner.run(testStepName);

        verify(testStepProcessor).process(testStep);
        verify(testScenarioState).set(testStep.getId(), testStep);
    }


    @Test
    @DisplayName("When step doesn't exist Then throw exception")
    void testRunTestStepWhenNotExist() throws URISyntaxException {
        initializeTestSteps("test1", Collections.emptyMap());
        initializeTestStepRunner();

        assertThrows(TestStepProcessingException.class, () -> testStepRunner.run("test2"));

        verifyZeroInteractions(testStepProcessor);
        verifyZeroInteractions(testScenarioState);
    }

    private void initializeTestSteps(String testStepName, Map<String, String> parameters) {
        Map<String, Object> rawTestStep = Collections.singletonMap("name", testStepName);

        testStep = new TestStep.Builder()
                .withId("t1")
                .withName(testStepName)
                .build();

        when(testStepFactory.create(testStepName, rawTestStep, parameters)).thenReturn(testStep);

        when(testStepReader.read(any(Path.class)))
                .thenReturn(Stream.of(rawTestStep)
                        .collect(Collectors.toList()));
    }

    private void initializeTestStepRunner() throws URISyntaxException {
        testStepRunner = new DefaultTestStepRunner.Builder()
                .withTestScenarioState(testScenarioState)
                .withTestStepFactory(testStepFactory)
                .withTestStepProcessor(testStepProcessor)
                .withTestStepReader(testStepReader, Collections.singletonList(Paths.get(getClass().getResource("/steps").toURI())))
                .build();
    }
}

File: skellig-test-step-runner/src/test/kotlin/org/skellig/teststep/runner/DefaultTestStepRunnerTest.kt
package org.skellig.teststep.runner

import org.mockito.kotlin.argThat
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.mockito.kotlin.doReturn
import org.skellig.teststep.processing.model.DefaultTestStep
import org.skellig.teststep.processing.model.TestStep
import org.skellig.teststep.processing.model.factory.TestStepFactory
import org.skellig.teststep.processing.processor.TestStepProcessor
import org.skellig.teststep.reader.TestStepReader
import org.skellig.teststep.reader.value.expression.AlphanumericValueExpression
import org.skellig.teststep.reader.value.expression.StringValueExpression
import org.skellig.teststep.reader.value.expression.ValueExpression
import org.skellig.teststep.runner.model.TestStepFileExtension
import org.skellig.teststep.runner.registry.TestStepsRegistry

@DisplayName("Run test step")
internal class DefaultTestStepRunnerTest {

    private var testStepRunner: TestStepRunner? = null
    private var testStep: TestStep? = null
    private var testStepProcessor = mock<TestStepProcessor<TestStep>>()
    private var testStepReader = mock<TestStepReader>()
    private var testStepFactory = mock<TestStepFactory<TestStep>>()

    @Test
    @DisplayName("When no parameters extracted from name")
    fun testRunTestStepWithoutParameters() {
        val testStepName = "test1"
        initializeTestSteps(testStepName, emptyMap<String, String>())
        initializeTestStepRunner()

        testStepRunner!!.run(testStepName)

        Mockito.verify(testStepProcessor).process(testStep!!)
    }

    @Test
    @DisplayName("When steps not found in the resources Then throw correct exception")
    fun testRunTestStepWhenNoFound() {
        val testStepName = "test1"
        initializeTestSteps(testStepName, emptyMap<String, String>())
        testStepRunner = DefaultTestStepRunner.Builder()
                .withTestStepFactory(testStepFactory)
                .withTestStepProcessor(testStepProcessor)
                .withTestStepsRegistry(createTestStepsRegistry("wrong path"))
                .build()

        Assertions.assertThrows(IllegalStateException::class.java) { testStepRunner!!.run(testStepName) }

        Mockito.verifyNoInteractions(testStepProcessor)
    }

    @Test
    @DisplayName("When step doesn't exist Then throw exception")
    fun testRunTestStepWhenNotExist() {
        initializeTestSteps("test1", emptyMap<String, String>())
        initializeTestStepRunner()

        Assertions.assertThrows(IllegalStateException::class.java) { testStepRunner!!.run("test2") }

        Mockito.verifyNoInteractions(testStepProcessor)
    }

    private fun initializeTestSteps(testStepName: String, parameters: Map<String, String?>) {
        val rawTestStep = mapOf<ValueExpression, ValueExpression?>(Pair(AlphanumericValueExpression("name"), StringValueExpression(testStepName)))
        testStep = DefaultTestStep.DefaultTestStepBuilder()
                .withId("t1")
                .withName(testStepName)
                .build()

        whenever(testStepFactory.create(testStepName, rawTestStep, parameters)).thenReturn(testStep)
        doReturn(listOf(rawTestStep)).whenever(testStepReader).read(argThat { true })
    }

    private fun initializeTestStepRunner() {
        testStepRunner = DefaultTestStepRunner.Builder()
                .withTestStepFactory(testStepFactory)
                .withTestStepProcessor(testStepProcessor)
                .withTestStepsRegistry(createTestStepsRegistry("steps"))
                .build()
    }

    private fun createTestStepsRegistry(testStepsPath: String): TestStepsRegistry {
        val testStepsRegistry = TestStepsRegistry(TestStepFileExtension.STS, testStepReader)
        val resource = javaClass.classLoader.getResource(testStepsPath)
        resource?.let {
            testStepsRegistry.registerFoundTestStepsInPath(listOf(it.toURI()))
        }
        return testStepsRegistry
    }
}